ktoml-core/src/commonMain/kotlin/com/akuleshov7/ktoml/encoders/TomlAbstractEncoder.kt

```diff
@@ -0,0 +1,388 @@
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.AbstractEncoder
+import kotlinx.serialization.modules.SerializersModule
+
+/**
+ * An abstract Encoder for the TOML format.
+ *
+ * @property elementIndex The current element index.
+ * @property attributes The current attributes.
+ * @property inputConfig The input config, used for constructing nodes.
+ * @property outputConfig The output config.
+ * @property serializersModule
+ */
+@OptIn(ExperimentalSerializationApi::class)
+public abstract class TomlAbstractEncoder protected constructor(
+    protected var elementIndex: Int,
+    protected val attributes: Attributes,
+    protected val inputConfig: TomlInputConfig,
+    protected val outputConfig: TomlOutputConfig,
+    override val serializersModule: SerializersModule,
+) : AbstractEncoder() {
+    private var isNextElementKey = false
+    private val instantDescriptor = Instant.serializer().descriptor
+    private val localDateTimeDescriptor = LocalDateTime.serializer().descriptor
+    private val localDateDescriptor = LocalDate.serializer().descriptor
+
+    protected open fun nextElementIndex(): Int = ++elementIndex
+
+    // Values
+
+    @Suppress("FUNCTION_BOOLEAN_PREFIX")
+    private fun encodeAsKey(key: Any, type: String? = null): Boolean {
+        if (!isNextElementKey) {
+            return false
+        }
+
+        isNextElementKey = false
+
+        if (key !is String) {
+            throw UnsupportedEncodingFeatureException(
+                "Arbitrary map key types are not supported. Must be either a string" +
+                        " or enum. Provide a custom serializer for $type to either " +
+                        "of the supported key types."
+            )
+        }
+
+        setKey(key)
+
+        return true
+    }
+
+    protected open fun appendValue(value: TomlValue) {
+        attributes.reset()
+    }
+
+    /**
+     * Allows [TomlInlineTableEncoder] and [TomlArrayEncoder] to access another
+     * encoder's [appendValue] function.
+     *
+     * @param value
+     * @param parent
+     */
+    internal fun appendValueTo(value: TomlValue, parent: TomlAbstractEncoder) {
+        parent.appendValue(value)
+    }
+
+    override fun encodeBoolean(value: Boolean) {
+        if (!encodeAsKey(value, "Boolean")) {
+            appendValue(TomlBoolean(value, elementIndex))
+        }
+    }
+
+    override fun encodeDouble(value: Double) {
+        if (!encodeAsKey(value, "Double")) {
+            appendValue(TomlDouble(value, elementIndex))
+        }
+    }
+
+    override fun encodeEnum(enumDescriptor: SerialDescriptor, index: Int) {
+        encodeString(enumDescriptor.getElementName(index))
+    }
+
+    override fun encodeLong(value: Long) {
+        if (attributes.intRepresentation != DECIMAL) {
+            throw UnsupportedEncodingFeatureException(
```

Comment:
in abstractDecoder we have a method `invalidType` may be make it common and use it here too?